File: bspline_fitting/Model/bspline_surface.py
import os
import torch
import numpy as np
import open3d as o3d
from typing import Union

import bs_fit_cpp

from bspline_fitting.Method.check import checkShape
from bspline_fitting.Method.value import toTorchPoints
from bspline_fitting.Method.render import renderPoints
from bspline_fitting.Method.path import createFileFolder, removeFile, renameFile


class BSplineSurface(object):
    def __init__(
        self,
        degree_u: int = 3,
        degree_v: int = 3,
        size_u: int = 5,
        size_v: int = 7,
        sample_num_u: int = 20,
        sample_num_v: int = 20,
        start_u: float = 0.0,
        start_v: float = 0.0,
        stop_u: float = 1.0,
        stop_v: float = 1.0,
        idx_dtype=torch.int64,
        dtype=torch.float64,
        device: str = "cpu",
    ) -> None:
        # Super Params
        self.degree_u = degree_u
        self.degree_v = degree_v
        self.size_u = size_u
        self.size_v = size_v
        self.sample_num_u = sample_num_u
        self.sample_num_v = sample_num_v
        self.start_u = start_u
        self.start_v = start_v
        self.stop_u = stop_u
        self.stop_v = stop_v
        self.idx_dtype = idx_dtype
        self.dtype = dtype
        self.device = device

        # Diff Params
        self.knotvector_u = torch.zeros(
            self.size_u - self.degree_u - 1, dtype=self.dtype
        ).to(self.device)
        self.knotvector_v = torch.zeros(
            self.size_v - self.degree_v - 1, dtype=self.dtype
        ).to(self.device)
        self.ctrlpts = torch.zeros(
            [(self.size_u - 1) * (self.size_v - 1), 3], dtype=self.dtype
        ).to(self.device)

        self.reset()
        return

    @classmethod
    def fromParamsDict(
        cls,
        params_dict: dict,
        sample_num_u: int = 20,
        sample_num_v: int = 20,
        start_u: float = 0.0,
        start_v: float = 0.0,
        stop_u: float = 1.0,
        stop_v: float = 1.0,
        idx_dtype=torch.int64,
        dtype=torch.float64,
        device: str = "cuda:0",
    ):
        degree_u = params_dict["degree_u"]
        degree_v = params_dict["degree_v"]
        size_u = params_dict["size_u"]
        size_v = params_dict["size_v"]

        bspline_surface = cls(
            degree_u,
            degree_v,
            size_u,
            size_v,
            sample_num_u,
            sample_num_v,
            start_u,
            start_v,
            stop_u,
            stop_v,
            idx_dtype,
            dtype,
            device,
        )

        bspline_surface.loadParamsDict(params_dict)

        return bspline_surface

    @classmethod
    def fromParamsFile(
        cls,
        params_file_path: str,
        sample_num_u: int = 20,
        sample_num_v: int = 20,
        start_u: float = 0.0,
        start_v: float = 0.0,
        stop_u: float = 1.0,
        stop_v: float = 1.0,
        idx_dtype=torch.int64,
        dtype=torch.float64,
        device: str = "cuda:0",
    ):
        params_dict = np.load(params_file_path, allow_pickle=True).item()

        return cls.fromParamsDict(
            params_dict,
            sample_num_u,
            sample_num_v,
            start_u,
            start_v,
            stop_u,
            stop_v,
            idx_dtype,
            dtype,
            device,
        )

    def reset(self) -> bool:
        self.initParams()
        return True

    def setGradState(self, need_grad: bool) -> bool:
        self.knotvector_u.requires_grad_(need_grad)
        self.knotvector_v.requires_grad_(need_grad)
        self.ctrlpts.requires_grad_(need_grad)
        return True

    def initParams(self) -> bool:
        return True

    def loadParams(
        self,
        knotvector_u: Union[torch.Tensor, np.ndarray, list, tuple, None] = None,
        knotvector_v: Union[torch.Tensor, np.ndarray, list, tuple, None] = None,
        ctrlpts: Union[torch.Tensor, np.ndarray, list, tuple, None] = None,
    ) -> bool:
        if knotvector_u is not None:
            if isinstance(knotvector_u, list) or isinstance(knotvector_u, tuple):
                knotvector_u = np.array(knotvector_u)

            if not checkShape(knotvector_u.shape, self.knotvector_u.shape):
                print("[ERROR][BSplineSurface::loadParams]")
                print("\t checkShape failed for knotvector_u!")
                return False

            if isinstance(knotvector_u, np.ndarray):
                knotvector_u = torch.from_numpy(knotvector_u)

            self.knotvector_u.data = (
                knotvector_u.detach().clone().type(self.dtype).to(self.device)
            )

        if knotvector_v is not None:
            if isinstance(knotvector_v, list) or isinstance(knotvector_v, tuple):
                knotvector_v = np.array(knotvector_v)

            if not checkShape(knotvector_v.shape, self.knotvector_v.shape):
                print("[ERROR][BSplineSurface::loadParams]")
                print("\t checkShape failed for knotvector_v!")
                return False

            if isinstance(knotvector_v, np.ndarray):
                knotvector_v = torch.from_numpy(knotvector_v)

            self.knotvector_v.data = (
                knotvector_v.detach().clone().type(self.dtype).to(self.device)
            )

        if ctrlpts is not None:
            if isinstance(ctrlpts, list) or isinstance(ctrlpts, tuple):
                ctrlpts = np.array(ctrlpts)

            if not checkShape(ctrlpts.shape, self.ctrlpts.shape):
                print("[ERROR][BSplineSurface::loadParams]")
                print("\t checkShape failed for ctrlpts!")
                return False

            if isinstance(ctrlpts, np.ndarray):
                ctrlpts = torch.from_numpy(ctrlpts)

            self.ctrlpts.data = (
                ctrlpts.detach().clone().type(self.dtype).to(self.device)
            )

        return True

    def loadParamsDict(self, params_dict: dict) -> bool:
        knotvector_u = params_dict["knotvector_u"]
        knotvector_v = params_dict["knotvector_v"]
        ctrlpts = params_dict["ctrlpts"]

        self.loadParams(knotvector_u, knotvector_v, ctrlpts)

        return True

    def loadParamsFile(self, params_file_path: str) -> bool:
        if not os.path.exists(params_file_path):
            print("[ERROR][BSplineSurface::loadParamsFile]")
            print("\t params dict file not exist!")
            print("\t params_file_path:", params_file_path)
            return False

        params_dict = np.load(params_file_path, allow_pickle=True).item()

        if not self.loadParamsDict(params_dict):
            print("[ERROR][BSplineSurface::loadParamsFile]")
            print("\t loadParamsDict failed!")
            return False

        return True

    def toSamplePoints(self) -> torch.Tensor:
        full_knotvector_u = torch.zeros(
            [self.degree_u + self.size_u], dtype=self.knotvector_u.dtype
        ).to(self.knotvector_u.device)
        full_knotvector_v = torch.zeros(
            [self.degree_v + self.size_v], dtype=self.knotvector_v.dtype
        ).to(self.knotvector_v.device)

        full_knotvector_u[-self.degree_u - 1 :] = 1.0
        full_knotvector_v[-self.degree_v - 1 :] = 1.0

        sigmoid_knotvector_u = torch.sigmoid(self.knotvector_u)
        sigmoid_knotvector_v = torch.sigmoid(self.knotvector_v)

        sigmoid_knotvector_u_sum = torch.sum(sigmoid_knotvector_u)
        sigmoid_knotvector_v_sum = torch.sum(sigmoid_knotvector_v)

        normed_sigmoid_knotvector_u = sigmoid_knotvector_u / sigmoid_knotvector_u_sum
        normed_sigmoid_knotvector_v = sigmoid_knotvector_v / sigmoid_knotvector_v_sum

        for i in range(normed_sigmoid_knotvector_u.shape[0] - 1):
            full_knotvector_u[self.degree_u + 1 + i] = (
                full_knotvector_u[self.degree_u + i] + normed_sigmoid_knotvector_u[i]
            )
        for i in range(normed_sigmoid_knotvector_v.shape[0] - 1):
            full_knotvector_v[self.degree_v + 1 + i] = (
                full_knotvector_v[self.degree_v + i] + normed_sigmoid_knotvector_v[i]
            )

        sample_points = toTorchPoints(
            self.degree_u,
            self.degree_v,
            self.size_u - 1,
            self.size_v - 1,
            self.sample_num_u,
            self.sample_num_v,
            self.start_u,
            self.start_v,
            self.stop_u,
            self.stop_v,
            full_knotvector_u,
            full_knotvector_v,
            self.ctrlpts,
        )

        return sample_points

    def renderSamplePoints(self) -> bool:
        sample_points = self.toSamplePoints().detach().clone().cpu().numpy()
        print(sample_points.shape)

        renderPoints(sample_points)
        return True

    def toParamsDict(self) -> dict:
        params_dict = {
            "degree_u": self.degree_u,
            "degree_v": self.degree_v,
            "size_u": self.size_u,
            "size_v": self.size_v,
            "knotvector_u": self.knotvector_u.detach().clone().cpu().numpy(),
            "knotvector_v": self.knotvector_v.detach().clone().cpu().numpy(),
            "ctrlpts": self.ctrlpts.detach().clone().cpu().numpy(),
        }
        return params_dict

    def saveParamsFile(
        self, save_params_file_path: str, overwrite: bool = False
    ) -> bool:
        if os.path.exists(save_params_file_path):
            if overwrite:
                removeFile(save_params_file_path)
            else:
                print("[WARN][BSplineSurface::saveParamsFile]")
                print("\t save params dict file already exist!")
                print("\t save_params_file_path:", save_params_file_path)
                return False

        params_dict = self.toParamsDict()

        createFileFolder(save_params_file_path)

        tmp_save_params_file_path = save_params_file_path[:-4] + "_tmp.npy"
        removeFile(tmp_save_params_file_path)

        np.save(tmp_save_params_file_path, params_dict)
        renameFile(tmp_save_params_file_path, save_params_file_path)
        return True

    def saveAsPcdFile(
        self,
        save_pcd_file_path: str,
        overwrite: bool = False,
        print_progress: bool = False,
        uniform_color=None,
    ) -> bool:
        if os.path.exists(save_pcd_file_path):
            if overwrite:
                removeFile(save_pcd_file_path)
            else:
                print("[ERROR][BSplineSurface::saveAsPcdFile]")
                print("\t save pcd file already exist!")
                print("\t save_pcd_file_path:", save_pcd_file_path)
                return False

        createFileFolder(save_pcd_file_path)

        points = self.toSamplePoints().detach().clone().cpu().numpy()

        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(points)
        if uniform_color is not None:
            pcd.paint_uniform_color(uniform_color)

        if print_progress:
            print("[INFO][BSplineSurface::saveAsPcdFile]")
            print("\t start save as pcd file...")
        o3d.io.write_point_cloud(
            save_pcd_file_path, pcd, write_ascii=True, print_progress=print_progress
        )
        return True

File: src/bspline_surface.cpp
#include "bspline_surface.h"
#include "value_torch.h"

using namespace torch::indexing;

BSplineSurface::BSplineSurface(const int &degree_u, const int &degree_v,
                               const int &size_u, const int &size_v,
                               const int &sample_num_u, const int &sample_num_v,
                               const float &start_u, const float &start_v,
                               const float &stop_u, const float &stop_v,
                               const std::string &idx_dtype,
                               const std::string &dtype,
                               const std::string &device) {
  is_valid_ = updateSetting(degree_u, degree_v, size_u, size_v, sample_num_u,
                            sample_num_v, start_u, start_v, stop_u, stop_v,
                            idx_dtype, dtype, device);

  if (!is_valid_) {
    std::cout << "[ERROR][BSplineSurface::BSplineSurface]" << std::endl;
    std::cout << "\t updateSetting failed!" << std::endl;

    return;
  }

  return;
}

const bool BSplineSurface::updateSetting(
    const int &degree_u, const int &degree_v, const int &size_u,
    const int &size_v, const int &sample_num_u, const int &sample_num_v,
    const float &start_u, const float &start_v, const float &stop_u,
    const float &stop_v, const std::string &idx_dtype, const std::string &dtype,
    const std::string &device) {
  degree_u_ = degree_u;
  degree_v_ = degree_v;
  size_u_ = size_u;
  size_v_ = size_v;
  sample_num_u_ = sample_num_u;
  sample_num_v_ = sample_num_v;
  start_u_ = start_u;
  start_v_ = start_v;
  stop_u_ = stop_u;
  stop_v_ = stop_v;

  if (idx_dtype == "int64") {
    idx_dtype_ = torch::kInt64;
  } else if (idx_dtype == "int32") {
    idx_dtype_ = torch::kInt32;
  } else {
    std::cout << "[ERROR][BSplineSurface::updateSetting]" << std::endl;
    std::cout << "\t idx dtype not valid!" << std::endl;

    return false;
  }

  if (dtype == "float64") {
    dtype_ = torch::kFloat64;
  } else if (dtype == "float32") {
    dtype_ = torch::kFloat32;
  } else {
    std::cout << "[ERROR][BSplineSurface::updateSetting]" << std::endl;
    std::cout << "\t dtype not valid!" << std::endl;

    return false;
  }

  if (device == "cpu") {
    device_ = torch::kCPU;
  } else if (device == "cuda") {
    device_ = torch::kCUDA;
  } else {
    std::cout << "[ERROR][BSplineSurface::updateSetting]" << std::endl;
    std::cout << "\t device not valid!" << std::endl;

    return false;
  }

  idx_opts_ = torch::TensorOptions().dtype(idx_dtype_).device(device_);
  opts_ = torch::TensorOptions().dtype(dtype_).device(device_);

  knotvector_u_ = torch::zeros({size_u_ - degree_u_ - 1}, opts_);
  knotvector_v_ = torch::zeros({size_v_ - degree_v_ - 1}, opts_);
  ctrlpts_ = torch::zeros({(size_u_ - 1) * (size_v_ - 1), 3}, opts_);

  return true;
}

const bool BSplineSurface::setGradState(const bool &need_grad) {
  knotvector_u_.set_requires_grad(need_grad);
  knotvector_v_.set_requires_grad(need_grad);
  ctrlpts_.set_requires_grad(need_grad);
  return true;
}

const bool BSplineSurface::loadCtrlPts(const torch::Tensor &ctrlpts) {
  ctrlpts_ = ctrlpts.detach().clone().toType(idx_dtype_).to(device_);
  return true;
}

const bool BSplineSurface::loadParams(const torch::Tensor &knotvector_u,
                                      const torch::Tensor &knotvector_v,
                                      const torch::Tensor &ctrlpts) {
  knotvector_u_ = knotvector_u.detach().clone().toType(idx_dtype_).to(device_);
  knotvector_v_ = knotvector_v.detach().clone().toType(idx_dtype_).to(device_);
  loadCtrlPts(ctrlpts);
  return true;
}

const torch::Tensor BSplineSurface::toSigmoidKnotvectorU() {
  torch::Tensor full_knotvector_u = torch::zeros({degree_u_ + size_u_}, opts_);

  full_knotvector_u.index_put_({Slice(size_u_ - 1, degree_u_ + size_u_)}, 1.0);

  const torch::Tensor sigmoid_knotvector_u = torch::sigmoid(knotvector_u_);

  const torch::Tensor sigmoid_knotvector_u_sum =
      torch::sum(sigmoid_knotvector_u);

  const torch::Tensor normed_sigmoid_knotvector_u =
      sigmoid_knotvector_u / sigmoid_knotvector_u_sum;

  for (int i = 0; i < normed_sigmoid_knotvector_u.size(0) - 1; ++i) {
    full_knotvector_u[degree_u_ + 1 + i] =
        full_knotvector_u[degree_u_ + i] + normed_sigmoid_knotvector_u[i];
  }

  return full_knotvector_u;
}

const torch::Tensor BSplineSurface::toSigmoidKnotvectorV() {
  torch::Tensor full_knotvector_v = torch::zeros({degree_v_ + size_v_}, opts_);

  full_knotvector_v.index_put_({Slice(size_v_ - 1, degree_v_ + size_v_)}, 1.0);

  const torch::Tensor sigmoid_knotvector_v = torch::sigmoid(knotvector_v_);

  const torch::Tensor sigmoid_knotvector_v_sum =
      torch::sum(sigmoid_knotvector_v);

  const torch::Tensor normed_sigmoid_knotvector_v =
      sigmoid_knotvector_v / sigmoid_knotvector_v_sum;

  for (int i = 0; i < normed_sigmoid_knotvector_v.size(0) - 1; ++i) {
    full_knotvector_v[degree_v_ + 1 + i] =
        full_knotvector_v[degree_v_ + i] + normed_sigmoid_knotvector_v[i];
  }

  return full_knotvector_v;
}

const torch::Tensor BSplineSurface::toSamplePoints() {
  const torch::Tensor sigmoid_knotvector_u = toSigmoidKnotvectorU();
  const torch::Tensor sigmoid_knotvector_v = toSigmoidKnotvectorV();

  const torch::Tensor sample_points = toTorchPoints(
      degree_u_, degree_v_, size_u_ - 1, size_v_ - 1, sample_num_u_,
      sample_num_v_, start_u_, start_v_, stop_u_, stop_v_, sigmoid_knotvector_u,
      sigmoid_knotvector_v, ctrlpts_);

  return sample_points;
}
